### PR TITLE
feat: Add kicbase download progress to JSON output

### LINKED ISSUE #15481: Add kicbase download progress to JSON output
The kicbase download doen't have JSON output steps, it should have them just like preload does.
```
{"specversion":"1.0","id":"1326f1ee-d21c-4c5f-8bbf-a54708514350","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"preloaded-images-k8s-v18-v1.25.3-docker-overlay2-arm64.tar.lz4","currentstep":"5","progress":"0.36961245448965924","totalsteps":"19"}}
{"specversion":"1.0","id":"2c5d9bc9-418a-48d6-b2be-17670a56a712","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"preloaded-images-k8s-v18-v1.25.3-docker-overlay2-arm64.tar.lz4","currentstep":"5","progress":"0.5720270019698556","totalsteps":"19"}}
{"specversion":"1.0","id":"10698714-7580-4144-9b3d-123c356666aa","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"preloaded-images-k8s-v18-v1.25.3-docker-overlay2-arm64.tar.lz4","currentstep":"5","progress":"0.7779482499838762","totalsteps":"19"}}
{"specversion":"1.0","id":"4fc8d74c-f26c-45ac-bcfd-47f5f8ee94f2","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"preloaded-images-k8s-v18-v1.25.3-docker-overlay2-arm64.tar.lz4","currentstep":"5","progress":"0.9248400390118532","totalsteps":"19"}}
{"specversion":"1.0","id":"3a4da4b1-1f50-4e81-84e4-70224beccf8c","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"preloaded-images-k8s-v18-v1.25.3-docker-overlay2-arm64.tar.lz4","currentstep":"5","progress":"1","totalsteps":"19"}}
{"specversion":"1.0","id":"7ed13212-9d20-4e84-bdc5-066a40fc08fc","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"preloaded-images-k8s-v18-v1.25.3-docker-overlay2-arm64.tar.lz4","currentstep":"5","progress":"1","totalsteps":"19"}}
```

### PR DESCRIPTION
**Pull request by ComradeProgrammer**
feat: Add kicbase download progress to JSON output

FIXES #15481 

**Before:**
Before this PR, when we execute `minikube start -o  json`, the output looks like this:
```
{"specversion":"1.0","id":"43c8e879-3385-44b0-83b6-127a6e19d18b","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.step","datacontenttype":"application/json","data":{"currentstep":"0","message":"minikube v1.28.0 on Ubuntu 22.04 (amd64)","name":"Initial Minikube Setup","totalsteps":"19"}}
{"specversion":"1.0","id":"4bb57aa3-c344-4235-bccc-0e29bb1461b2","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.step","datacontenttype":"application/json","data":{"currentstep":"1","message":"Automatically selected the docker driver","name":"Selecting Driver","totalsteps":"19"}}
{"specversion":"1.0","id":"20590f8a-5be6-47ab-8ab8-f625a9044921","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.info","datacontenttype":"application/json","data":{"message":"Using image repository registry.cn-hangzhou.aliyuncs.com/google_containers"}}
{"specversion":"1.0","id":"364a2ec2-8cf7-44a5-a42b-8e9b7f059c8d","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.info","datacontenttype":"application/json","data":{"message":"Using Docker driver with root privileges"}}
{"specversion":"1.0","id":"9f721f95-1a70-4e46-b812-7eaf4b7e9165","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.step","datacontenttype":"application/json","data":{"currentstep":"3","message":"Starting control plane node minikube in cluster minikube","name":"Starting Node","totalsteps":"19"}}
{"specversion":"1.0","id":"eff70d70-6c03-4634-b6dc-39bd010411be","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.step","datacontenttype":"application/json","data":{"currentstep":"5","message":"Pulling base image ...","name":"Pulling Base Image","totalsteps":"19"}}
{"specversion":"1.0","id":"a66b877b-0aa2-4a01-9ce3-a3ed85b49220","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.step","datacontenttype":"application/json","data":{"currentstep":"8","message":"Creating docker container (CPUs=2, Memory=2200MB) ...","name":"Creating Container","totalsteps":"19"}}
......
```
No output for kicbase download progress in json output
**After**
After this PR, when we execute `minikube start -o  json`, we can find this part
```
......(similar logs at the beginning are omitted) 

{"specversion":"1.0","id":"ba07f967-db4b-4684-8d6b-6d613dbbfb0e","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"docker.io/kicbase/build:v0.0.36-1673540226-15630@sha256:03c9592728381094cbd0ff9603f75ae6b485dd7a390c3e35f02ae5ec10f2f3ad","currentstep":"5","progress":"0","totalsteps":"19"}}
{"specversion":"1.0","id":"a1f4296c-47f7-4983-bed2-3d2b2842bd85","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"docker.io/kicbase/build:v0.0.36-1673540226-15630@sha256:03c9592728381094cbd0ff9603f75ae6b485dd7a390c3e35f02ae5ec10f2f3ad","currentstep":"5","progress":"0.000013","totalsteps":"19"}}

......(lines of logs omitted)

{"specversion":"1.0","id":"f32754ae-0733-4cd8-9772-0dc9bf354e89","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"docker.io/kicbase/build:v0.0.36-1673540226-15630@sha256:03c9592728381094cbd0ff9603f75ae6b485dd7a390c3e35f02ae5ec10f2f3ad","currentstep":"5","progress":"0.999978","totalsteps":"19"}}
{"specversion":"1.0","id":"cbcd8e85-d102-4a4f-9c88-91b1f1ab13a4","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"docker.io/kicbase/build:v0.0.36-1673540226-15630@sha256:03c9592728381094cbd0ff9603f75ae6b485dd7a390c3e35f02ae5ec10f2f3ad","currentstep":"5","progress":"0.999981","totalsteps":"19"}}
{"specversion":"1.0","id":"8be461b8-1d3c-4919-920f-24a88f0030d9","source":"https://minikube.sigs.k8s.io/","type":"io.k8s.sigs.minikube.download.progress","datacontenttype":"application/json","data":{"artifact":"docker.io/kicbase/build:v0.0.36-1673540226-15630@sha256:03c9592728381094cbd0ff9603f75ae6b485dd7a390c3e35f02ae5ec10f2f3ad","currentstep":"5","progress":"1","totalsteps":"19"}}
......(similar logs at the end are omitted) 
```

Similar to downloading progress for preloads, the output is genearted per second.

### CODE REVIEW

**Comment by ComradeProgrammer** [comment]
@spowelljr

**Review by medyagh — COMMENTED** [review]
I am curious how many new lines does this add to the json file ? maybe the json users wouldnt need that level of details

**Comment by ComradeProgrammer** [comment]
> I am curious how many new lines does this add to the json file ? maybe the json users wouldnt need that level of details

Well, for normal users with a good network connection(10MB/S), there would be about 48 lines?. Kicbase image is 480MB.

If that is too much, I can make it generate less logs.

**Review by spowelljr — CHANGES_REQUESTED** [review]
I believe this outputs a JSON log every second, for people with slow internet the download can take a long time and the JSON output can get quite large. I'd recommend this outputs say every 5 seconds instead to lessen the logs while still giving consistent output that the process is still running.

**Comment by ComradeProgrammer** [comment]
> I believe this outputs a JSON log every second, for people with slow internet the download can take a long time and the JSON output can get quite large. I'd recommend this outputs say every 5 seconds instead to lessen the logs while still giving consistent output that the process is still running.

(Well the current code is exactly  logging every 5 seconds

File: pkg/minikube/download/image.go [hunk 1]
```diff
@@ -139,38 +138,54 @@ func ImageToCache(img string) error {
...
+	errchan := make(chan error)
+	// buffered channel
+	c := make(chan v1.Update, 200)
+	var p *pb.ProgressBar
 	if out.JSON {
-		if err := tarball.WriteToFile(f, tag, i); err != nil {
-			return errors.Wrap(err, "writing tarball image")
+		register.PrintDownloadProgress(img, "0")
+	} else {
+		// f we need to print progress bat to stdout
```
Comment (spowelljr) [inline pkg/minikube/download/image.go:148]:
```suggestion
		// if we need to print progress bar to stdout
```
Not sure if `f` is supposed to be `if` or not?

**Review by spowelljr — CHANGES_REQUESTED** [review]

**Pull request merged**